### DISCUSSION TITLE
8088 support

### DISCUSSION THREAD

**Discussion by darvin** [open-watcom/open-watcom-v2#1106] [Ideas]
need 8088 support to run it under [MartyPC](https://github.com/dbalsom/martypc)

**Comment by joshudson** [comment]
Hypothetically I've been wondering about this myself. OpenWatcom is a compiler that can target 8086 processor, and hosts itself. It seems that it should work to tell it to build itself targeting 8086 but probably has issues. It's just a question of how many.

**Comment by winspool** [comment]
Running OpenWatcom on DOS needs a 32bit DOS extender (included in OpenWatcom)
so a 386 processor is the minimum.


You can use OpenWatcom to cross compile your programs
for a 16 bit DOS target (8086, 80186 or 80286) 
and when your source needs floating point support,
using 8087, 80287 and floating point software emulation is possible.


Running your compiled programs on a 8088 (with the 8 bit data bus)
should work, but that is slower than running on the 8086 (16 bit data bus).

**Comment by joshudson** [comment]
@winspool : And if the program being cross-compiled is OpenWatcom itself?

**Reply by davidrg** [reply]
According to MichalN on the openwatcom IRC several days back, Watcom 9.0 (1992) was the last version to include 16-bit compilers.

Given no one has built a 16-bit version of Watcom in over 30 years I'm guessing it may not be easy.

**Reply by jmalak** [reply]
Open Watom Compilers source code is not compilable as 16-bit executable, only as 32-bit.
You can not get Watcom 9.x source code because OW is based on Watcom 11.x version which was only 32-bit executable.

**Comment by winspool** [comment]
> And if the program being cross-compiled is OpenWatcom itself?

The 640k memory limit of DOS and the segment size of 64k are the problems here.

To work with a resonable amount of memory,
the application needs code to work around the limits.
One or both of EMS or XMS might be available, but both are complicated and slow.

Much simpler solution to handle the memory requirement of a compiler toolchain
is using the 32bit protected mode of the processor
and use the DPMI programming interface for calling back to 16bit DOS and 16bit BIOS.
That solution is also used in OpenWatcom, when running under DOS.

**Comment by joshudson** [comment]
I was suspecting the solution to the memory problem was disable whole program optimization and compile each function independently. But anyway, it's clearly a lot of work.

I recommend just leaving this item open and seeing just how much desire really does show up. Most likely not enough.
